gateway/pastaporto/__tests__/identity.test.ts

```diff
@@ -0,0 +1,291 @@
+
+    expect(() => decodeRefreshToken(testToken, "20")).toThrowWithMessage(
+      JsonWebTokenError,
+      "jwt subject invalid. expected: 20",
+    );
+  });
+
+  test("Reject expired refresh token", () => {
+    jest
+      .useFakeTimers("modern")
+      .setSystemTime(new Date("2050-03-20 14:40:31Z").getTime());
+
+    const testToken =
+      "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2MTYyNTA5MzEsImV4cCI6MTYyMzY4MTMzMSwiYXVkIjoicmVmcmVzaCIsImlzcyI6ImdhdGV3YXkiLCJzdWIiOiIxMCJ9.14tohtJ9NfLFQ0wq6tKVHEC4Opphh37Z5xABsy7XVP8";
+
+    expect(() => decodeRefreshToken(testToken, "10")).toThrow(
+      TokenExpiredError,
+    );
+  });
+
+  test("Reject refresh token with wrong audience", () => {
+    jest
+      .useFakeTimers("modern")
+      .setSystemTime(new Date("2021-03-20 14:40:31Z").getTime());
+
+    const testToken =
+      "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2MTYyNTA5MzEsImV4cCI6MTYyMzY4MTMzMSwiYXVkIjoiaWRlbnRpdHkiLCJpc3MiOiJnYXRld2F5Iiwic3ViIjoiMTAifQ.UQfAUpUc2h-Sbjl59or2cCSUQJ9fG0Vsky4HwbOtAhA";
+
+    expect(() => decodeRefreshToken(testToken, "10")).toThrowWithMessage(
+      JsonWebTokenError,
+      "jwt audience invalid. expected: refresh",
+    );
+  });
+
+  test("Reject refresh token with wrong issuer", () => {
+    jest
+      .useFakeTimers("modern")
+      .setSystemTime(new Date("2021-03-20 14:40:31Z").getTime());
+
+    const testToken =
+      "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2MTYyNTA5MzEsImV4cCI6MTYyMzY4MTMzMSwiYXVkIjoicmVmcmVzaCIsImlzcyI6InVzZXJzIiwic3ViIjoiMTAifQ.Aw3FjtRtp50eeus_zrvLFoPpjJRM-wV5sRdrd3sYRPk";
+
+    expect(() => decodeRefreshToken(testToken, "10")).toThrowWithMessage(
+      JsonWebTokenError,
+      "jwt issuer invalid. expected: gateway",
+    );
+  });
+
+  test("Reject refresh token with wrong secret key", () => {
+    jest
+      .useFakeTimers("modern")
+      .setSystemTime(new Date("2021-03-20 14:40:31Z").getTime());
+
+    // signed with abc123
+    const testToken =
+      "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2MTYyNTA5MzEsImV4cCI6MTYyMzY4MTMzMSwiYXVkIjoicmVmcmVzaCIsImlzcyI6ImdhdGV3YXkiLCJzdWIiOiIxMCJ9.NZX6RETwxM4R0cBV5_Ubjo98gTeubPArTVKgs1BJpH8";
+
+    expect(() => decodeRefreshToken(testToken, "10")).toThrowWithMessage(
+      JsonWebTokenError,
+      "invalid signature",
+    );
+  });
+});
+
+describe("Create refresh token", () => {
+  test("Create refresh token", () => {
+    jest
+      .useFakeTimers("modern")
+      .setSystemTime(new Date("2021-03-20 14:35:31Z").getTime());
+
+    const expectedToken =
+      "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2MTYyNTA5MzEsImV4cCI6MTYyMzUwODUzMSwiYXVkIjoicmVmcmVzaCIsImlzcyI6ImdhdGV3YXkiLCJzdWIiOiIxMCJ9.8nF91uPV7ngknINxD0X9U58t-Z7JPwogQw9ngx6cNI4";
+
+    expect(createRefreshToken("10")).toBe(expectedToken);
+  });
+
+  test("Cannot create refresh token with empty subject", () => {
+    expect(() => createRefreshToken("")).toThrow("Empty subject not allowed");
+  });
+});
```

Comment:
awesome work, wouldn't it be best to use an utility function to generate these tokens? they are hard to scan/debug 😊

Reply:
sure, will change. I guess we can assume that the utility that creates the token is always right, one of the reason I used pre-created tokens was so we would be safe from bugs in the jwt library itself

Reply:
done